webhookValidator.go

```diff
@@ -70,145 +70,75 @@ func (vf ValidatorFunc) Validate(w Webhook) error {
 	return vf(w)
 }
 
-// filterNil takes out all entries of Nil value from the slice.
-func filterNil(vs []ValidURLFunc) (filtered []ValidURLFunc) {
-	for _, v := range vs {
-		if v != nil {
-			filtered = append(filtered, v)
-		}
-	}
-	return
-}
-
-// GoodConfigURL parses the given webhook's Config.URL
-// and returns as soon as the URL is considered invalid. It returns nil if the URL is
-// valid.
-func GoodConfigURL(vs []ValidURLFunc) ValidatorFunc {
-	vs = filterNil(vs)
+// CheckEvents makes sure there is at least one value in Events and ensures that
+// all values should parse into regex.
+func CheckEvents() ValidatorFunc {
 	return func(w Webhook) error {
-		parsedURL, err := url.ParseRequestURI(w.Config.URL)
-		if err != nil {
-			return fmt.Errorf("%w: %v", errInvalidURL, err)
+		if len(w.Events) == 0 {
+			return errZeroEvents
 		}
-		for _, f := range vs {
-			err = f(parsedURL)
+		for _, e := range w.Events {
+			_, err := regexp.Compile(e)
 			if err != nil {
-				return fmt.Errorf("%w: %v", errInvalidURL, err)
+				return errEventsUnparseable
 			}
 		}
 		return nil
 	}
 }
 
-// GoodFailureURL parses the given webhook's FailureURL
-// and returns as soon as the URL is considered invalid. It returns nil if the URL is
-// valid.
-func GoodFailureURL(vs []ValidURLFunc) ValidatorFunc {
-	vs = filterNil(vs)
+// CheckDeviceID ensures that the DeviceIDs are able to parse into regex.
+func CheckDeviceID() ValidatorFunc {
 	return func(w Webhook) error {
-		if w.FailureURL != "" {
-			parsedFailureURL, err := url.ParseRequestURI(w.FailureURL)
+		for _, i := range w.Matcher.DeviceID {
+			_, err := regexp.Compile(i)
 			if err != nil {
-				return fmt.Errorf("%w: %v", errInvalidFailureURL, err)
-			}
-			for _, f := range vs {
-				err = f(parsedFailureURL)
-				if err != nil {
-					return fmt.Errorf("%w: %v", errInvalidFailureURL, err)
-				}
+				return errDeviceIDUnparseable
 			}
 		}
 		return nil
 	}
 }
 
-// GoodAlternativeURLs parses the given webhook's Config.AlternativeURLs
-// and returns as soon as the URL is considered invalid. It returns nil if the URL is
-// valid.
-func GoodAlternativeURLs(vs []ValidURLFunc) ValidatorFunc {
-	vs = filterNil(vs)
+// CheckDuration ensures that 0 <= Duration <= ttl. Duration returns an error
+// if a negative value is given.
+func CheckDuration(ttl time.Duration) ValidatorFunc {
 	return func(w Webhook) error {
-		for _, u := range w.Config.AlternativeURLs {
-			if u == "" {
-				return errInvalidAlternativeURL
-			}
-			parsedAlternativeURL, err := url.ParseRequestURI(u)
-			if err != nil {
-				return fmt.Errorf("%w: %v", errInvalidAlternativeURL, err)
-			}
-			for _, f := range vs {
-				err = f(parsedAlternativeURL)
-				if err != nil {
-					return fmt.Errorf("%w: %v", errInvalidAlternativeURL, err)
-				}
-			}
+		if ttl < 0 {
```

Comment:
This is a good thing to check!

Instead of running this check on each validator function call, we should check the ttl before returning a `ValidatorFunc`.  `CheckDuration()`'s function signature can return `(ValidatorFunc, error)` and the caller of the function won't receive a `ValidatorFunc` at all if the ttl is bad.

This allows us to know the ttl is bad at startup rather than the application constantly erroring out while running.